r/tests/testthat/test-backwards-compatibility.R

```diff
@@ -0,0 +1,171 @@
+  expect_equal(
+    attributes(df$c),
+    list(
+      row.names = 1L,
+      names = c("c1", "c2", "c3"),
+      class = c("tbl_df", "tbl", "data.frame")
+    )
+  )
+  expect_null(attributes(df$d))
+})
+
+for (comp in c("lz4", "uncompressed", "zstd")) {
+  # write_feather(example_with_metadata, test_path("golden-files/data-arrow_2.0.0_lz4.feather"), compression = "lz4")
+  # write_feather(example_with_metadata, test_path("golden-files/data-arrow_2.0.0_uncompressed.feather"), compression = "uncompressed")
+  # write_feather(example_with_metadata, test_path("golden-files/data-arrow_2.0.0_zstd.feather"), compression = "zstd")
+  test_that("reading a known Feather file to dataframe with 2.0.0", {
+    skip_if_not_available(comp)
+    feather_file <- test_path(paste0("golden-files/data-arrow_2.0.0_", comp,".feather"))
+
+    df <- read_feather(feather_file)
+    expect_equal(df, example_with_metadata)
+    expect_identical(dim(df), c(1L, 4L))
+
+    expect_equal(
+      attributes(df),
+      list(
+        names = letters[1:4],
+        row.names = 1L,
+        top_level = list(field_one = 12, field_two = "more stuff"),
+        class = c("tbl_df", "tbl", "data.frame"))
+    )
+    expect_equal(attributes(df$a), list(class = "special_string"))
+    expect_null(attributes(df$b))
+    expect_equal(
+      attributes(df$c),
+      list(
+        row.names = 1L,
+        names = c("c1", "c2", "c3"),
+        class = c("tbl_df", "tbl", "data.frame")
+      )
+    )
+    expect_null(attributes(df$d))
+  })
+
+  test_that("reading a known Feather file to dataframe with 1.0.1", {
+    skip_if_not_available(comp)
+    feather_file <- test_path(paste0("golden-files/data-arrow_1.0.1_", comp,".feather"))
+
+    df <- read_feather(feather_file)
+    # 1.0.1 didn't save top-level metadata, so we need to remove it.
+    example_with_metadata_sans_toplevel <- example_with_metadata
+    attributes(example_with_metadata_sans_toplevel)$top_level <- NULL
+    expect_equal(df, example_with_metadata_sans_toplevel)
+    expect_identical(dim(df), c(1L, 4L))
+
+    expect_equal(
+      attributes(df),
+      list(
+        names = letters[1:4],
+        row.names = 1L,
+        class = c("tbl_df", "tbl", "data.frame"))
+    )
+    expect_equal(attributes(df$a), list(class = "special_string"))
+    expect_null(attributes(df$b))
+    expect_equal(
+      attributes(df$c),
+      list(
+        row.names = 1L,
+        names = c("c1", "c2", "c3"),
+        class = c("tbl_df", "tbl", "data.frame")
+      )
+    )
+    expect_null(attributes(df$d))
+  })
+
+  test_that("reading a known Feather file to dataframe with 0.17.0", {
+    skip_if_not_available(comp)
+    feather_file <- test_path(paste0("golden-files/data-arrow_0.17.0_", comp,".feather"))
+
+    if (comp %in% c("lz4", "zstd")) {
+      # there is a case mis-match with versions 0.17.0 and before for the codec names
+      expect_error(df <- read_feather(feather_file), "Unrecognized compression type:")
```

Comment:
So once https://github.com/apache/arrow/pull/9128 is merged, this should be fixed

Reply:
Great, I'll change these to skips and than add the Jira to the message then.